gtsam/geometry/Similarity2.cpp

```diff
@@ -0,0 +1,203 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file   Similarity2.cpp
+ * @brief  Implementation of Similarity2 transform
+ * @author John Lambert
+ */
+
+#include <gtsam/geometry/Similarity2.h>
+
+#include <gtsam/geometry/Pose2.h>
+#include <gtsam/base/Manifold.h>
+#include <gtsam/slam/KarcherMeanFactor-inl.h>
+
+namespace gtsam {
+
+using std::vector;
+
+namespace {
+/// Subtract centroids from point pairs.
+static Point2Pairs subtractCentroids(const Point2Pairs &abPointPairs,
+                                    const Point2Pair &centroids) {
+  Point2Pairs d_abPointPairs;
+  for (const Point2Pair& abPair : abPointPairs) {
+    Point2 da = abPair.first - centroids.first;
+    Point2 db = abPair.second - centroids.second;
+    d_abPointPairs.emplace_back(da, db);
+  }
+  return d_abPointPairs;
+}
+
+/// Form inner products x and y and calculate scale.
+static const double calculateScale(const Point2Pairs &d_abPointPairs,
+                                   const Rot2 &aRb) {
+  double x = 0, y = 0;
```

Comment:
Move da, db declarations into loop, make refs?

Reply:
Didn't quite follow the comment here.

Reply:
```
{
  Point2 &da, &db;
  std::tie(da, db) = d_abPair;
...
}
```

NOT for this PR but I now think we should have a Matrix version of Rot2, Pose2, and and do
```
DA = A.colwise() - A.mean();  // 2xN
DB = B.colwise() - B.mean();  // 2xN
aRb * DB * DA.transpose();  //  2x2 * 2xN * Nx2 = 2x2
// I think x and y should be found in the matrix above?
```

Reply:
Even I don't understand the purpose of this. It's doable but why?

Reply:
And because it's a reference, initialization of the variables is required, which means the above 2 lines will change to the single line

```cpp
const Point2 &da = d_abPair.first, &db = d_abPair.second;
```